File: backend/internal/domain/usecases/category_usecase.go
package usecases

import (
	"catalogo-produtos/backend/internal/domain/entities"
	"catalogo-produtos/backend/internal/domain/repositories"
	"errors"
)

// CategoryUseCase define os casos de uso para categorias
type CategoryUseCase interface {
	CreateCategory(name string) (*entities.Category, error)
	GetCategory(id uint) (*entities.Category, error)
	GetCategories() ([]entities.Category, error)
	UpdateCategory(id uint, name string) (*entities.Category, error)
	DeleteCategory(id uint) error
}

// categoryUseCase implementa CategoryUseCase
type categoryUseCase struct {
	categoryRepo repositories.CategoryRepository
}

// NewCategoryUseCase cria uma nova instância de CategoryUseCase
func NewCategoryUseCase(categoryRepo repositories.CategoryRepository) CategoryUseCase {
	return &categoryUseCase{
		categoryRepo: categoryRepo,
	}
}

// CreateCategory cria uma nova categoria
func (uc *categoryUseCase) CreateCategory(name string) (*entities.Category, error) {
	// Validar nome
	if name == "" {
		return nil, errors.New("nome é obrigatório")
	}

	category := &entities.Category{
		Name: name,
	}

	err := uc.categoryRepo.Create(category)
	if err != nil {
		return nil, err
	}

	return category, nil
}

// GetCategory busca uma categoria por ID
func (uc *categoryUseCase) GetCategory(id uint) (*entities.Category, error) {
	category, err := uc.categoryRepo.GetByID(id)
	if err != nil {
		return nil, errors.New("categoria não encontrada")
	}
	return category, nil
}

// GetCategories busca todas as categorias
func (uc *categoryUseCase) GetCategories() ([]entities.Category, error) {
	categories, err := uc.categoryRepo.GetAll()
	if err != nil {
		return nil, err
	}
	return categories, nil
}

// UpdateCategory atualiza uma categoria
func (uc *categoryUseCase) UpdateCategory(id uint, name string) (*entities.Category, error) {
	// Buscar categoria existente
	category, err := uc.categoryRepo.GetByID(id)
	if err != nil {
		return nil, errors.New("categoria não encontrada")
	}

	// Validar nome
	if name == "" {
		return nil, errors.New("nome é obrigatório")
	}

	// Atualizar nome
	category.Name = name

	err = uc.categoryRepo.Update(category)
	if err != nil {
		return nil, err
	}

	return category, nil
}

// DeleteCategory remove uma categoria
func (uc *categoryUseCase) DeleteCategory(id uint) error {
	// Verificar se a categoria existe
	_, err := uc.categoryRepo.GetByID(id)
	if err != nil {
		return errors.New("categoria não encontrada")
	}

	return uc.categoryRepo.Delete(id)
}

File: frontend/src/domain/usecases/CategoryUseCase.ts
import { Category } from '../entities/Product';
import { CategoryRepository } from '../repositories/CategoryRepository';

// Casos de uso para categorias
export interface CategoryUseCase {
  getCategories(): Promise<string[]>;
  getCategory(id: number): Promise<Category>;
  getCategoryByName(name: string): Promise<Category | null>;
  createCategory(name: string): Promise<Category>;
  updateCategory(id: number, name: string): Promise<Category>;
  deleteCategory(id: number): Promise<void>;
}

// Implementação dos casos de uso
export class CategoryUseCaseImpl implements CategoryUseCase {
  constructor(private categoryRepository: CategoryRepository) {}

  async getCategories(): Promise<string[]> {
    try {
      return await this.categoryRepository.getCategories();
    } catch (error) {
      console.error('Erro ao buscar categorias:', error);
      throw new Error('Falha ao carregar categorias');
    }
  }

  async getCategory(id: number): Promise<Category> {
    try {
      return await this.categoryRepository.getCategory(id);
    } catch (error) {
      console.error('Erro ao buscar categoria:', error);
      throw new Error('Categoria não encontrada');
    }
  }

  async getCategoryByName(name: string): Promise<Category | null> {
    try {
      return await this.categoryRepository.getCategoryByName(name);
    } catch (error) {
      console.error('Erro ao buscar categoria por nome:', error);
      throw new Error('Falha ao buscar categoria');
    }
  }

  async createCategory(name: string): Promise<Category> {
    try {
      // Validações de negócio
      if (!name || name.trim().length === 0) {
        throw new Error('Nome da categoria é obrigatório');
      }
      if (name.trim().length < 2) {
        throw new Error('Nome da categoria deve ter pelo menos 2 caracteres');
      }

      return await this.categoryRepository.createCategory(name.trim());
    } catch (error) {
      console.error('Erro ao criar categoria:', error);
      throw error;
    }
  }

  async updateCategory(id: number, name: string): Promise<Category> {
    try {
      // Validações de negócio
      if (!name || name.trim().length === 0) {
        throw new Error('Nome da categoria é obrigatório');
      }
      if (name.trim().length < 2) {
        throw new Error('Nome da categoria deve ter pelo menos 2 caracteres');
      }

      return await this.categoryRepository.updateCategory(id, name.trim());
    } catch (error) {
      console.error('Erro ao atualizar categoria:', error);
      throw error;
    }
  }

  async deleteCategory(id: number): Promise<void> {
    try {
      await this.categoryRepository.deleteCategory(id);
    } catch (error) {
      console.error('Erro ao deletar categoria:', error);
      throw new Error('Falha ao deletar categoria');
    }
  }
} 